### ISSUE TITLE
Indexes should start at 0

### ISSUE THREAD

**Issue by arturock** [PHPOffice/PhpSpreadsheet#279]
I'm confused with the decision about using 1 as starting index.
It's dev counter-intuitive.
Consistency with Excel seem practically irrelevant as that's an end user app, not a programming library.
I'm aware that the parent package works that way but I would argue that a bad practice from 7 years ago should not be inherited.
A serious package should stick to standard common practices on its exposed API whenever possible.
It would be pretty weird for new comers to use indexes that way because a package they never used before had that particularity.
[Why numbering should start at zero](http://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html) by Dijkstra may worth a read

**Comment by PowerKiKi** [comment]
So are you arguing that saying "row 0 in code will end up as row 1 in Excel" is less confusing ?

In any other contexts I would absolutely go with zero based indexes. But here we have a strong precedent which is also reflected in the file format for XLSX where column indexes start at 1 (so they did change from XLS starting at 0). And while PhpSpreadsheet is indeed intended for developers, they will still have to reason about Excel files as a "normal" person would. So sticking to whatever Excel does, makes it easier to explain and most likely more familiar for most people who ever used Excel.

Also I find Dijkstra justification quite convoluted. And honestly I am not sure it even applies here. He seems very focused on how to express ranges. But those are much less relevant in modern language where `foreach` would be the de-facto loop. And even with `for` loop you would very seldom express range directly, instead writing something like `for ($i = 1; $i <= $max; $i++)`. I honestly don't see how it can be objectively argued to be harder to read/write than the more traditional `for ($i = 0; $i < $max; $i++)`. It makes very little difference to me. But being able to say "indexes are just like in Excel, everything starts at 1" is valuable, because it is very easy to explain, consistent between rows and columns, consistent between app and lib, and easily understandable.

I'll be closing this issue for now, because it is unlikely to be changed at this point (but not impossible). However the discussion can keep going...

**Comment by arturock** [comment]
> So are you arguing that saying "row 0 in code will end up as row 1 in Excel" is less confusing ?

As the question (maybe unintentionally) suggests that those two values should be the same, let me rephrase it to an even more basic form:

So are you arguing that saying "row 0 in code will end up as first row in Excel" is less confusing ?

Yes, that's what I'd expected. Having to use 1, -1 or whatever for accessing first one is a collateral effect (from the programmer point of view) when using the app. As I said, when coding, the person is not the Excel end-user but the dev, those are different roles so they shouldn't be mixed even when playing both.
I totally concur with XLSX on using 1, as its one of the reasons I'm suggesting here to use 0.

I do see the reasoning behind the current decision and their benefits but I just don't feel it as the way to go, however I'm not a heavy Excel user myself.

Keep up the great work!

**Comment by PowerKiKi** [comment]
Indeed "row 0 in code will end up as first row in Excel" is extremely easy to understand and logical for developers.

But, from experience, that changes when you are not talking about "first row" anymore, but instead about any other arbitrary row. When you are looking at a spreadsheet opened in Excel and thinking something along the lines of:

> I need to read those values between rows 43 and 54 and columns A to D, but I need to write down 42 and 53 for rows in my code, because they are shifted, but 0 to 4 for columns, because columns don't need to be shifted... or is that the other way around ?

It can be unexpectedly confusing to have to wonder everytime you use a column or row if they must be shifted or not. The way it is now, you learn it once "all indexes always starts at 1", or alternatively "exactly like Excel" and you're done.
